### PR TITLE
Fix tar extraction in mingw windows

### PR DESCRIPTION
**Pull request by calebkiage**
MinGW has a unix compatible tar program on windows. The tar program from the msys registry doesn't have a bundled 7zip extractor.

* Before adding new features and new modules, please go to issues to submit the relevant feature description first.
* Write good commit messages and use the same coding conventions as the rest of the project.
* Please commit code to dev branch and we will merge into master branch in feature
* Ensure your edited codes with four spaces instead of TAB.

------

* 增加新特性和新模块之前，请先到issues提交相关特性说明，经过讨论评估确认后，再进行相应的代码提交，避免做无用工作。
* 编写友好可读的提交信息，并使用与工程代码相同的代码规范，代码请用4个空格字符代替tab缩进。
* 请提交代码到dev分支，如果通过，我们会在特定时间合并到master分支上。
* 为了规范化提交日志的格式，commit消息，不要用中文，请用英文描述。

### CODE REVIEW

File: xmake/modules/utils/archive/extract.lua [hunk 1]
```diff
@@ -33,7 +33,7 @@ import("extension", {alias = "get_archive_extension"})
 function _extract_using_tar(archivefile, outputdir, extension, opt)
 
     -- the tar of windows can only extract "*.tar"
-    if os.host() == "windows" and extension ~= ".tar" then
+    if os.host() == "windows" and not os.is_subhost("msys", "cygwin") and extension ~= ".tar" then
```
Comment (waruqi) [inline xmake/modules/utils/archive/extract.lua:36]: no, msys2/cygwin's tar does not decompress many non-.tar files. So I disabled it as well. Had a lot of problems with it before

File: xmake/core/base/os.lua [hunk 2]
```diff
@@ -1002,8 +1002,9 @@ end
 -- get the system null device
 function os.nuldev(input)
 
+    -- mingw and cygwin have /dev/null
     if input then
-        if os.host() == "windows" then
+        if os.host() == "windows" and not os.is_subhost("msys", "cygwin") then
```
Comment (waruqi) [inline xmake/core/base/os.lua:1007]: I know msys2 has /dev/null. But it seems to have had some problems before. The compatibility wasn't very good, so I disabled it as well.

Comment (waruqi) [reply xmake/core/base/os.lua:1007]: Also, xmake's os.exec is always a direct call to CreateProcess, which doesn't support /dev/null for msys2. xmake doesn't use mingw's fork to create processes.

File: xmake/modules/utils/archive/extract.lua (hunk 1 above)
Comment (calebkiage) [reply xmake/modules/utils/archive/extract.lua:36]: Ok. I'll remove this

File: xmake/core/base/os.lua (hunk 2 above)
Comment (calebkiage) [reply xmake/core/base/os.lua:1007]: I'll remove this too.

**Pull request merged**